packages/core/lib/patchers/aws3_p.ts

```diff
@@ -0,0 +1,179 @@
+import type {
+  MetadataBearer,
+  ResponseMetadata,
+  Command,
+  Client,
+} from '@aws-sdk/types';
+
+import type { RegionInputConfig } from '@aws-sdk/config-resolver';
+
+import { isThrottlingError } from '@aws-sdk/service-error-classification';
+
+import ServiceSegment from '../segments/attributes/aws';
+
+import { stringify } from 'querystring';
+
+import {
+  Subsegment,
+  Segment,
+  getNamespace,
+  resolveSegment,
+  setSegment,
+} from '..';
+
+
+var contextUtils = require('../context_utils');
+
+var logger = require('../logger');
+
+import { getCauseTypeFromHttpStatus } from '../utils';
+
+type HttpResponse = { response: { status: number } };
+
+async function buildAttributesFromMetadata(
+  client: any,
+  command: any,
+  metadata: ResponseMetadata,
+): Promise<[ServiceSegment, HttpResponse]> {
+  const { httpHeaders, httpStatusCode: statusCode, retries } = metadata;
+  const serviceIdentifier = client.config.signingName as string;
+
+  let operation: string = command.constructor.name.slice(0, -7);
+  operation = operation.charAt(0).toLowerCase() + operation.slice(1);
+
+  const aws = new ServiceSegment(
+    {
+      extendedRequestId: httpHeaders && httpHeaders['x-amz-id-2'],
+      requestId: httpHeaders && httpHeaders['x-amz-request-id'],
+      retryCount: retries,
+      request: {
+        operation,
+        httpRequest: {
+          region: await client.config.region(),
+          statusCode,
+        },
+        params: command.input,
+      },
+    },
+    serviceIdentifier,
+  );
+
+  const http = { response: { status: statusCode || 0 } };
+  return [aws, http];
+}
+
+function addFlags(http: HttpResponse, subsegment: Subsegment, err?: Error): void {
+  if (err && isThrottlingError(err)) {
+    subsegment.addThrottleFlag();
+  } else if (http.response?.status === 429) {
+    subsegment.addThrottleFlag();
+  }
+
+  const cause = getCauseTypeFromHttpStatus(http.response.status);
+  if (cause === 'fault') {
+    subsegment.addFaultFlag();
+  } else if (cause === 'error') {
+    subsegment.addErrorFlag();
+  }
+}
+
+type DefaultConfiguration = RegionInputConfig & {
+  signingName: string;
+};
+
+export function captureAWSClient<
+  Input extends object,
+  Output extends MetadataBearer,
+  Configuration extends DefaultConfiguration
+>(client: Client<Input, Output, Configuration>): Client<Input, Output, Configuration> {
+  // create local copy so that we can later call it
+  const send = client.send;
+
+  const serviceIdentifier = client.config.signingName;
+
+  client.send = async (command: any): Promise<Output> => {
+    const segment = contextUtils.resolveSegment();
```

Comment:
Yeah, so manual mode will be tough in TypeScript. Is there any way that, when calling `captureAWSClient`, we could iterate over all its methods and add an additional optional argument for the `XRaySegment`? Or could we add it as a field to the `Configuration` object?

Reply:
`captureAWSClient` now takes a second, optional segment argument. That would address the `manual` mode. This means that for manual mode, you need to call `captureAWSClient` any time you create a new segment. That way we can ensure that we don't break the TypeScript experience for any developers.

Reply:
Hmmm this would be a pretty big break from the other types of manual mode instrumentation we have - calling `captureAWSClient` is generally a one-and-done deal. But then again it is called manual mode for a reason, I'll discuss this with the team as well.